core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/inverserelation/ExternalizedSingletonInverseVariableSupply.java

```diff
@@ -32,8 +32,7 @@ public VariableDescriptor<Solution_> getSourceVariableDescriptor() {
     @Override
     public void resetWorkingSolution(ScoreDirector<Solution_> scoreDirector) {
         inverseEntityMap = new IdentityHashMap<>();
-        sourceVariableDescriptor.getEntityDescriptor().getSolutionDescriptor()
-                .visitAllEntities(scoreDirector.getWorkingSolution(), this::insert);
+        sourceVariableDescriptor.getEntityDescriptor().visitAllEntities(scoreDirector.getWorkingSolution(), this::insert);
```

Comment:
Part of the fix.

Context: a separate CH phase is automatically created for each genuine planning entity with the chained variable. The CH phase uses a `ChangeMoveSelector` and that demands a `SingletonInverseVariableSupply` (implemented by this class).

So there is one `ChangeMoveSelector` for `Enity1.chainedVar1` and another one for `Entity2.chainedVar2` and each has its own `ExternalizedSingletonInverseVariableSupply` with a `sourceVariableDescriptor` based on one of {chainedVar1, chainedVar2}.

Now, if on _this line_, before the fix, _all entities_ are visited (so both Entity1 and Entity2 instances), then this supply's `this::insert` method will call `sourceVariableDescriptor.getValue(entity)` with entity being an instance of the other class than the one that owns the source variable.